File: LeetCode/00448. Find All Numbers Disappeared in an Array.py
#Time Limit Exceeded
class Solution:
    def findDisappearedNumbers(self, nums):
        r = []
        for index in range(1, len(nums)+1):
            if index not in nums:
                r.append(index)
        return r

# Accepted
class Solution:
    def findDisappearedNumbers(self, nums):
        e = set(range(1, len(nums)+ 1))
        for num in nums:
            if num in e: e.remove(num)
        return list(e)
      
      
# Another Solution
class Solution:
    def findDisappearedNumbers(self, nums):
        sorted_nums = sorted(nums)
        n = max(nums) + 1
        e = set(range(1, len(nums)+1)).difference(sorted_nums)
        return list(e)

File: LeetCode/00448. Find All Numbers Disappeared in an Array.java
// using hashset
class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
      List<Integer> output = new ArrayList<>();
      Set<Integer> set = new HashSet<>();
      
      for (int num : nums) {
        set.add(num);
      }
      
      int n = nums.length;
      for (int i = 1; i <= n; i++) {
        if (!set.contains(i)) {
          output.add(i);
        }
      }
        return output;
    }
}


// another solution using set
class Solution {
  public List<Integer> findDisappearedNumbers(int[] nums) {
    Set<Integer> expectedNumbers = new HashSet<>(range(1, nums.length));
    for (int num : nums) expectedNumbers.remove(num);
    return new ArrayList<>(expectedNumbers);
  }

  private static List<Integer> range(int start, int end) {
    List<Integer> list = new ArrayList<>();
    for (int i = start; i <= end; i++) list.add(i);
    return list;
  }
}
